sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/PathFilterSuite.scala

```diff
@@ -0,0 +1,557 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources
+
+import java.io.File
+import java.time.LocalDateTime
+
+import org.apache.hadoop.fs.Path
+
+import org.apache.spark.sql.{AnalysisException, QueryTest, Row}
+import org.apache.spark.sql.catalyst.util.{stringToFile, CaseInsensitiveMap, DateTimeUtils}
+import org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanHelper
+import org.apache.spark.sql.execution.datasources.pathfilters.PathFilterFactory
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.test.SharedSparkSession
+
+class PathFilterSuite
+    extends QueryTest
+    with SharedSparkSession
+    with AdaptiveSparkPlanHelper {
+  import testImplicits._
+
+  test("SPARK-31962: when modifiedAfter specified with a past date") {
+    withTempDir { dir =>
+      withSQLConf(SQLConf.SESSION_LOCAL_TIMEZONE.key -> "UTC") {
+        val path = new Path(dir.getCanonicalPath)
+        val file = new File(dir, "file1.csv")
+        stringToFile(file, "text")
+        file.setLastModified(DateTimeUtils.currentTimestamp())
+        val df = spark.read
+          .option("modifiedAfter", "2019-05-10T01:11:00")
+          .format("csv")
+          .load(path.toString)
+        assert(df.count() == 1)
+      }
+    }
+  }
+
+  test("SPARK-31962: when modifiedBefore specified with a future date") {
+    withTempDir { dir =>
+      withSQLConf(SQLConf.SESSION_LOCAL_TIMEZONE.key -> "UTC") {
+        val path = new Path(dir.getCanonicalPath)
+        val file = new File(dir, "file1.csv")
+        stringToFile(file, "text")
+        val df = spark.read
+          .option("modifiedBefore", "2090-05-10T01:11:00")
+          .format("csv")
+          .load(path.toString)
+        assert(df.count() == 1)
+      }
+    }
+  }
+
+ test("SPARK-31962: when modifiedBefore specified with a past date") {
+    withTempDir { dir =>
+      withSQLConf(SQLConf.SESSION_LOCAL_TIMEZONE.key -> "UTC") {
+        val path = new Path(dir.getCanonicalPath)
+        val file = new File(dir, "file1.csv")
+        stringToFile(file, "text")
+        file.setLastModified(DateTimeUtils.currentTimestamp())
+        val msg = intercept[AnalysisException] {
+          spark.read
+            .option("modifiedBefore", "1984-05-01T01:00:00")
+            .format("csv")
+            .load(path.toString)
+        }.getMessage
+        assert(msg.contains("Unable to infer schema for CSV"))
```

Comment:
What's the purpose of this test? `Unable to infer schema for CSV`?

Reply:
When attempting to load a CSV-based file data source, if no files are returned, this is the error message returned.  It was existing default behavior.  Whenever a filter scenario results in all files being filtered, this exception is caught and used to identify the result.